storage/memory.js

```diff
@@ -0,0 +1,321 @@
+    if (existingKey) {
+      oldReferences = this.keysReferences.get(key)
+      this.log.debug({ msg: 'acd/storage/memory.set, current keys-references', key, references: oldReferences })
+      if (oldReferences) {
+        oldReferences.sort()
+        references.sort()
+        const referencesToRemove = findNotMatching(references, oldReferences)
+
+        // remove key in old references
+        for (const reference of referencesToRemove) {
+          const keys = this.referencesKeys.get(reference)
+          // istanbul ignore next
+          if (!keys) { continue }
+          const index = bsearchIndex(keys, key)
+          // istanbul ignore next
+          if (index < 0) { continue }
+          keys.splice(index, 1)
+
+          if (keys.length < 1) {
+            this.referencesKeys.delete(reference)
+            continue
+          }
+          this.referencesKeys.set(reference, keys)
+        }
+      }
+    }
+
+    // TODO we can probably get referencesToAdd and referencesToRemove in a single loop
+    const referencesToAdd = oldReferences ? findNotMatching(oldReferences, references) : references
+
+    for (let i = 0; i < referencesToAdd.length; i++) {
+      const reference = referencesToAdd[i]
+      let keys = this.referencesKeys.get(reference)
+      if (keys) {
+        this.log.debug({ msg: 'acd/storage/memory.set, add reference-key', key, reference })
+        keys.push(key)
+        keys.sort()
+      } else {
+        keys = [key]
+      }
+      this.log.debug({ msg: 'acd/storage/memory.set, set reference-keys', keys, reference })
+      this.referencesKeys.set(reference, keys)
+    }
+
+    this.keysReferences.set(key, references)
+  }
+
+  /**
+   * remove an entry by key
+   * @param {string} key
+   * @returns {boolean} indicates if key was removed
+   */
+  async remove (key) {
+    this.log.debug({ msg: 'acd/storage/memory.remove', key })
+
+    const removed = this._removeKey(key)
+    this._removeReferences([key])
+    return removed
+  }
+
+  /**
+   * @param {string} key
+   * @returns {boolean}
+   */
+  _removeKey (key) {
+    this.log.debug({ msg: 'acd/storage/memory._removeKey', key })
+    if (!this.store.has(key)) {
+      return false
+    }
+    this.store.set(key, undefined)
+    return true
+  }
+
+  /**
+   * @param {string[]} keys
+   */
+  _removeReferences (keys) {
+    this.log.debug({ msg: 'acd/storage/memory._removeReferences', keys })
+
+    const referencesToRemove = new Set()
+    for (let i = 0; i < keys.length; i++) {
+      const key = keys[i]
+
+      const references = this.keysReferences.get(key)
+      if (!references) {
+        continue
+      }
+
+      for (let j = 0; j < references.length; j++) {
+        referencesToRemove.add(references[j])
+      }
+
+      this.log.debug({ msg: 'acd/storage/memory._removeReferences, delete key-references', key })
+      this.keysReferences.delete(key)
+    }
+
+    this._removeReferencesKeys([...referencesToRemove], keys)
```

Comment:
why converting this to an array?

Reply:
I used the set to avoid repetitions, but keys are arrays because I need to sort them